src/lib.rs

```diff
@@ -32,9 +32,42 @@
 mod windows;
 
 use std::env;
+use std::ffi::OsString;
 use std::io;
 use std::path::{Path, PathBuf};
 
+/// Permits parameterising the home functions via the _from variants - used for
+/// in-process unit testing by rustup.
+pub trait Env {
+    fn home_dir(&self) -> Option<PathBuf>;
+    fn current_dir(&self) -> io::Result<PathBuf>;
+    fn var_os(&self, key: &str) -> Option<OsString>;
+}
+
+/// Implements Env for the OS context, both Unix style and Windows.
+///
+/// This is trait permits in-process testing by providing a control point to
+/// allow in-process divergence on what is normally process wide state.
+///
+/// Implementations should be provided by whatever testing framework the caller
+/// is using. Code that is not performing in-process threaded testing requiring
+/// isolated rustup/cargo directories does not need this trait or the _from
+/// functions.
+pub struct OsEnv {}
+impl Env for OsEnv {
+    fn home_dir(&self) -> Option<PathBuf> {
+        home_dir_inner()
+    }
+    fn current_dir(&self) -> io::Result<PathBuf> {
+        env::current_dir()
+    }
+    fn var_os(&self, key: &str) -> Option<OsString> {
+        env::var_os(key)
+    }
+}
+
+pub static OS_ENV: OsEnv = OsEnv {};
```

Comment:
This is zero-sized struct? I think make this const item make more sense?
Also if we has default implementation for the trait, this struct is needed to be public.

Reply:
const should be fine, I can do that too. See my other comment about publicness.

Reply:
One reason I prefer const struct is that it allow optimizing deref to OS_ENV.